src/prog_models/models/uav_model/uav_model.py

```diff
@@ -0,0 +1,556 @@
+            dist_now = np.sqrt((x['x']-x_next)**2 + (x['y']-y_next)**2 + (x['z']-z_next)**2)
+            if dist_now <= self.parameters['final_space_buffer_m']:
+                # Waypoint achieved
+                self.parameters['waypoints']['next_waypoint'] += 1
+                return {
+                    'TrajectoryComplete': (num_wypts - index_next)/num_wypts
+                }
+            else:
+                # Waypoint not yet achieved
+                return {
+                    'TrajectoryComplete': (num_wypts - (index_next - 1))/num_wypts
+                }
+        else:
+            # ETA passed before waypoint reached 
+            warn("Trajectory may not have reached waypoint associated with ETA of {})".format(t_next))
+            self.parameters['waypoints']['next_waypoint'] += 1
+            return {
+                    'TrajectoryComplete': (num_wypts - index_next)/num_wypts
+                }
+ 
+    def output(self, x : dict):
+        # Output is the same as the state vector
+        return self.OutputContainer({
+            'x': x['x'],
+            'y': x['y'],
+            'z': x['z'],
+            'phi': x['phi'],
+            'theta': x['theta'],
+            'psi': x['psi'],
+            'vx': x['vx'],
+            'vy': x['vy'],
+            'vz': x['vz'],
+            'p': x['p'],
+            'q': x['q'],
+            'r': x['r'],
+            't': x['t']
+            })
+
+    def threshold_met(self, x : dict) -> dict:
+        t_lower_bound = self.parameters['waypoints']['waypoints_time'][-1] - (self.parameters['waypoints']['waypoints_time'][-1] - self.parameters['waypoints']['waypoints_time'][-2])/2
+        t_upper_bound = self.parameters['waypoints']['waypoints_time'][-1] + self.parameters['final_time_buffer_sec']
+        if x['t'] < t_lower_bound:
+            # Trajectory hasn't reached final ETA
+            return {
+                'TrajectoryComplete': False
+            }
+        elif t_lower_bound <= x['t'] <= t_upper_bound:
+            # Trajectory is within bounds of final ETA
+            dist_now = np.sqrt((x['x']-self.parameters['waypoints']['waypoints_x'][-1])**2 + (x['y']-self.parameters['waypoints']['waypoints_y'][-1])**2 + (x['z']-self.parameters['waypoints']['waypoints_z'][-1])**2)
+            if dist_now <= self.parameters['final_space_buffer_m']:
+                return {
+                    'TrajectoryComplete': True
+                }
+            else: 
+                return {
+                    'TrajectoryComplete': False
+                }
+        else: 
+            # Trajectory has passed acceptable bounds of final ETA - simulation terminated
+            warn("Trajectory simulation extends beyond the final ETA. Either the final waypoint was not reached in the alotted time (and the simulation was terminated), or simulation was run for longer than the trajectory length.")
```

Comment:
What if the final waypoint hasn't been met yet, but we're running behind? Is that a case you would want to support

Reply:
It will continue simulating for a buffer time defined by the parameter "final_time_buffer_sec" past the final ETA time. Originally I implemented it this way because I was thinking that if the final waypoint wasn't met, the simulation would go on forever. But now I'm realizing that the default horizon would eventually stop this. However, once we pass the final ETA, we won't have any new input information, so the results quickly level out to whatever output corresponds to that input. Does that make sense? I'm happy to leave it as is or get rid of this and let it run to horizon

Reply:
OK- I see why you implemented it that way it is. If you're simulating to threshold wouldn't it stop anyways at the last waypoint, so it wouldn't simulate past the last input information. 

I guess that doesn't help for simulate to. I'm hoping you can remove this when we move to the modular, since input would be provided by the future loading equation instead. I don't like the idea of having two horizons